src/Marten/Events/Daemon/Coordination/ProjectionCoordinator.cs

```diff
@@ -0,0 +1,252 @@
+using System.Threading;
+using System.Threading.Tasks;
+using JasperFx.Core;
+using Marten.Events.Daemon.Resiliency;
+using Marten.Storage;
+using Microsoft.Extensions.Logging;
+using Polly;
+
+namespace Marten.Events.Daemon.Coordination;
+
+public class ProjectionCoordinator<T>: ProjectionCoordinator, IProjectionCoordinator<T> where T : IDocumentStore
+{
+    public ProjectionCoordinator(T documentStore, ILogger<ProjectionCoordinator> logger) : base(documentStore, logger)
+    {
+    }
+}
+
+public class ProjectionCoordinator : IProjectionCoordinator
+{
+    private readonly StoreOptions _options;
+    private readonly ILogger<ProjectionCoordinator> _logger;
+
+    private readonly ConcurrentDictionary<IMartenDatabase, IProjectionDaemon> _daemons = new();
+    private readonly ResiliencePipeline _resilience;
+    private readonly CancellationTokenSource _cancellation = new();
+    private Task _runner;
+
+    public ProjectionCoordinator(IDocumentStore documentStore, ILogger<ProjectionCoordinator> logger)
+    {
+        var store = (DocumentStore)documentStore;
+
+        if (store.Options.Projections.AsyncMode == DaemonMode.Solo)
+        {
+            Distributor = new SoloProjectionDistributor(store);
+        }
+        else if (store.Options.Projections.AsyncMode == DaemonMode.HotCold)
+        {
+            if (store.Options.Tenancy is DefaultTenancy)
+            {
+                Distributor = new SingleTenantProjectionDistributor(store);
+            }
+            else
+            {
+                Distributor = new MultiTenantedProjectionDistributor(store);
+            }
+        }
+
+        _options = store.Options;
+        _logger = logger;
+        _resilience = store.Options.ResiliencePipeline;
+
+        Store = store;
+    }
+
+    public IProjectionDaemon DaemonForMainDatabase()
+    {
+        var database = (MartenDatabase)Store.Tenancy.Default.Database;
+        if (_daemons.TryGetValue(database, out var daemon))
+        {
+            return daemon;
+        }
+
+        daemon = database.StartProjectionDaemon(Store, _logger);
+        _daemons[database] = daemon;
+
+        return daemon;
+    }
+
+    public async ValueTask<IProjectionDaemon> DaemonForDatabase(string databaseIdentifier)
+    {
+        var database = (MartenDatabase)await Store.Storage.FindOrCreateDatabase(databaseIdentifier).ConfigureAwait(false);
+        if (_daemons.TryGetValue(database, out var daemon))
+        {
+            return daemon;
+        }
+
+        daemon = database.StartProjectionDaemon(Store, _logger);
+        _daemons[database] = daemon;
+
+        return daemon;
+    }
+
+    public DocumentStore Store { get; }
+
+    public IProjectionDistributor Distributor { get; }
+
+    internal record DaemonShardName(IProjectionDaemon Daemon, ShardName Name);
+
+    public Task StartAsync(CancellationToken cancellationToken)
+    {
+        _runner = Task.Run(() => executeAsync(_cancellation.Token), _cancellation.Token);
```

Comment:
**CONCERN:** Why are we adding a `Task.Run` wrapper here if executeAsync already returns a task?

Reply:
I understand, but why do we need `Task.Run` wrapper instead of just running `executeAsync`?